### DISCUSSION TITLE
Parsing golang depdency version list to exclude identified versions

### DISCUSSION THREAD

**Discussion by ajhcpt** [renovatebot/renovate#17747] [Request Help]
Hi, 

I am trying renovate to update golang dependencies. We proxy through Artifactory for all our dependencies. For our external depedencies this is working fine. The private depedency resolution is not working and its seems like its due to how artifactory displays the version for these dependencies:

https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/list

> v0.0.1 2022-03-07T01:41:41Z
v1.0.0 2021-07-07T06:05:27Z
v1.0.1 2021-07-07T01:08:02Z
v1.0.2 2021-08-03T08:36:21Z
v1.0.3 2021-08-03T11:18:24Z
v1.0.4 2021-08-03T12:01:04Z
v1.0.5 2021-08-04T06:18:51Z
v1.0.6 2021-08-04T11:52:46Z

The date seems to be an issue since renovate seems to identify v0.0.1 and "2022-03-07T01:41:41Z" as separate versions.

> [90mTRACE[39m: Can't obtain data from https://private.artificatory.test.net/artifactory/api/go/go-remote-repo[90m (repository=test/api-go-renovate-test)[39m
       "err": {
         "name": "HTTPError",
         "code": "ERR_NON_2XX_3XX_RESPONSE",
         "timings": {
           "start": 1662967610796,
           "socket": 1662967610797,
           "lookup": 1662967610842,
           "connect": 1662967610857,
           "secureConnect": 1662967610869,
           "upload": 1662967610872,
           "response": 1662967611158,
           "end": 1662967611158,
           "phases": {
             "wait": 1,
             "dns": 45,
             "tcp": 15,
             "tls": 12,
             "request": 3,
             "firstByte": 286,
             "download": 0,
             "total": 362
           }
         },
         "message": "Response code 404 (Not Found)",
         "stack": "HTTPError: Response code 404 (Not Found)\n    at Request.<anonymous> (/usr/src/app/node_modules/got/dist/source/as-promise/index.js:118:42)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)",
         "options": {
           "headers": {
             "user-agent": "RenovateBot/32.168.0 (https://github.com/renovatebot/renovate)",
             "accept": "application/json",
             "accept-encoding": "gzip, deflate, br"
           },
           "url": "https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/2022-03-07T01:41:41Z.info",
           "hostType": "go-proxy",
           "username": "",
           "password": "",
           "method": "GET",
           "http2": false
         },
         "response": {
           "statusCode": 404,
           "statusMessage": "Not Found",
           "body": {"errors": [{"status": 404, "message": "Not Found"}]},
           "headers": {
             "date": "Mon, 12 Sep 2022 07:26:50 GMT",
             "content-type": "application/json",
             "transfer-encoding": "chunked",
             "connection": "close",
             "x-jfrog-version": "Artifactory/7.35.2 73502900",
             "x-artifactory-node-id": "test-artifactory-01"
           },
           "httpVersion": "1.1"
         }
       }

JFrog has been very unhelpful in figuring out why the dates are displayed along with the versions for our internally published go artifacts. Given that, I've was hoping to use renovate configuration to ignore the date "versions" being identified. I've tried following with variations of the allowedVersions regex  (where all our versions are of the pattern v1.0.0) without any success:

`  {
  matchManagers: ['gomod'],
  matchPackagePrefixes: ['test.net'],
  allowedVersions: '/^v[0-9]+\\.[0-9]+\\.[0-9]+/',
}`

Is my regex incorrect or is the allowedVersions only applied after attempting to reach the .info urls for each dependency? Is there any other way to exclude the tool from identifiying the dates in the list output as a version?

**Comment by PhilipAbed** [comment]
step 1: extract dependencies gomod
step 2: look up the versions available
step 3: check if the allowed versions are part of the available released versions that returned from registry

the URL that we are calling is wrong in step 2 
url:`https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/2022-03-07T01:41:41Z.info`

so questions for you:
- what is the actual URL we should be calling?
- does your private repo need a token to connect to? are you using host rules to do that?
- according to the above url the extracted version seems to be `2022-03-07T01:41:41Z` so probably we dont extract the right version for your dependencies , 
did you try working around it using [regexManagers](https://docs.renovatebot.com/configuration-options/#regexmanagers) instead?

**Reply by ajhcpt** [reply]
I see that this is where the go dependency versions are processed:

https://github.com/renovatebot/renovate/blob/main/lib/modules/datasource/go/releases-goproxy.ts#L201

Would it be possible to change this to something like:

```  
  async listVersions(baseUrl: string, packageName: string): Promise<string[]> {
    const url = `${baseUrl}/${this.encodeCase(packageName)}/@v/list`;
    const { body } = await this.http.get(url);
    return body
      .split(regEx(/\n/))
      .map(version => version.split(regEx(/\s+/))[0])
      .filter(Boolean)
      .filter((x) => x.indexOf('+') === -1);
  }
```

**Comment by ajhcpt** [comment]
Thanks for the response, hopefully I can explain a bit better:

From the logs I can see renovate calling this URL, which is the correct url to get the versions for the "test.net/validation" dependency:

https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/list

The result from this call is:
> v0.0.1 2022-03-07T01:41:41Z
v1.0.0 2021-07-07T06:05:27Z
v1.0.1 2021-07-07T01:08:02Z
v1.0.2 2021-08-03T08:36:21Z
v1.0.3 2021-08-03T11:18:24Z
v1.0.4 2021-08-03T12:01:04Z
v1.0.5 2021-08-04T06:18:51Z
v1.0.6 2021-08-04T11:52:46Z

where normal go dependencies would look like this:
> v0.0.1
v1.0.0
v1.0.1
v1.0.2
v1.0.3
v1.0.4
v1.0.5
v1.0.6

From the logs it looks like renovate parses this response to check for each version and then makes a call for each.

So it starts with:

https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v0.0.1.info

which is correct. 

but it is seeing `v0.0.1 2022-03-07T01:41:41Z` as two separate versions:
- v0.0.1
- 2022-03-07T01:41:41Z

and so then attempts to hit this URL:

https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/2022-03-07T01:41:41Z.info

which results in a 404 since that is not a valid version. 

The only urls I want to be checked are:
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v0.0.1.info
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v1.0.0.info
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v1.0.1.info
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v1.0.2.info
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v1.0.3.info
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v1.0.4.info
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v1.0.5.info
https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/v1.0.6.info

To answer your questions:
- what is the actual URL we should be calling? the correct URLs are being called for vX.X.X, but would like to prevent the "2022-..." date portions from being identified as a version.
- does your private repo need a token to connect to? are you using host rules to do that? I'm running the renovate docker container and have loaded our certificates into the container, so no token is required.
- using RegexManagers: correct me if I'm wrong, but from what I can tell in the documentation this is used to parse files (i.e uses the fileMatch param) in the repository and would not work for the response from a version call like https://private.artificatory.test.net/artifactory/api/go/go-remote-repo/test.net/validation/@v/list.

**Reply by rarkins** [reply]
@zharinov sounds like our parsing could be tightened up?

**Reply by zharinov** [reply]
Taking a look

**Reply by zharinov** [reply]
Fix PR is done: https://github.com/renovatebot/renovate/pull/17785

**Reply by ajhcpt** [reply]
wow! thanks for the quick turnaround, i'll try out the updated release.
